gremlin-go/driver/driverRemoteConnection.go

```diff
@@ -80,14 +85,38 @@ func NewDriverRemoteConnection(
 		transporterType: settings.TransporterType,
 		logHandler:      logHandler,
 		connection:      connection,
+		session:         settings.Session,
+		closed:          false,
 	}
 
 	return &DriverRemoteConnection{client: client}, nil
 }
 
 // Close closes the DriverRemoteConnection.
 func (driver *DriverRemoteConnection) Close() error {
-	return driver.client.Close()
+	// If DriverRemoteConnection has spawnedSessions then they must be closed as well.
+	if len(driver.spawnedSessions) > 0 {
+		driver.client.logHandler.logger.Logf(Info, "closing spawned sessions from DriverRemoteConnection with url '%s'", driver.client.url)
+		for _, session := range driver.spawnedSessions {
+			err := session.Close()
+			if err != nil {
+				return err
+			}
+		}
+		driver.spawnedSessions = []*DriverRemoteConnection{}
+	}
+
+	if driver.IsSessionBound() {
+		driver.client.logHandler.logger.Logf(Info, "closing DriverRemoteConnection with url '%s' with session '%s'", driver.client.url, driver.client.session)
+	} else {
+		driver.client.logHandler.logger.Logf(Info, "closing DriverRemoteConnection with url '%s'", driver.client.url)
+	}
+	err := driver.client.Close()
+	if err != nil {
+		return err
+	}
+	driver.client.closed = true
```

Comment:
I don't think you should modify the state tracker of a different type outside of the type.